test/smtchecker.js

```diff
@@ -1,25 +1,107 @@
 
   t.test('smoke test with no smtlib2queries', function (st) {
     const input = {};
     const output = { auxiliaryInputRequested: {} };
-    st.equal(smtchecker.handleSMTQueries(input, output), null);
+    st.equal(smtchecker.handleSMTQueries(input, output, null, null), null);
     st.end();
   });
 
   t.test('smoke test with empty smtlib2queries', function (st) {
     const input = {};
     const output = { auxiliaryInputRequested: { smtlib2queries: { } } };
-    st.equal(smtchecker.handleSMTQueries(input, output), null);
+    st.equal(smtchecker.handleSMTQueries(input, output, null, null), null);
+    st.end();
+  });
+
+  t.test('smtCallback should return type function', (st) => {
+    const response = smtchecker.smtCallback(() => {});
+    st.equal(typeof response, 'function');
+    st.end();
+  });
+
+  t.test('smtCallback should error when passed parser fails', (st) => {
+    const callback = smtchecker.smtCallback((content) => { throw new Error(content); });
+    // eslint-disable-next-line node/no-callback-literal
+    const response = callback('expected-error-message');
+
+    st.deepEqual(response, { error: new Error('expected-error-message') });
+    st.end();
+  });
+
+  t.test('smtCallback should return content when passed parser does not fail', (st) => {
+    const callback = smtchecker.smtCallback((content) => { return content; });
+    // eslint-disable-next-line node/no-callback-literal
+    const response = callback('expected-content-message');
+
+    st.deepEqual(response, { contents: 'expected-content-message' });
+    st.end();
+  });
+});
+
+tape('SMTCheckerWithSolver', function (t) {
+  // In these tests we require z3 to actually run the solver.
+  // This uses the SMT double run mechanism instead of the callback.
+
+  t.test('Simple test with axuiliaryInputRequested', function (st) {
+    const z3 = smtsolver.availableSolvers.filter(solver => solver.command === 'z3');
+    if (z3.length === 0) {
+      st.skip('Test requires z3.');
+      st.end();
+      return;
+    }
+
+    if (semver.lt(solc.semver(), '0.8.7')) {
+      st.skip('This test requires Solidity 0.8.7 to enable all SMTChecker options.');
+      st.end();
+      return;
+    }
+
+    const settings = {
+      modelChecker: {
+        engine: 'chc',
+        solvers: ['smtlib2']
+      }
+    };
+
+    const source = { a: { content: preamble + '\ncontract C { function f(uint x) public pure { assert(x > 0); } }' } };
+
+    const input = {
+      language: 'Solidity',
+      sources: source,
+      settings: settings
+    };
+
+    const output = JSON.parse(solc.compile(JSON.stringify(input)));
+    st.ok(output);
+
+    const newInput = smtchecker.handleSMTQueries(input, output, smtsolver.smtSolver, z3[0]);
+    st.notEqual(newInput, null);
+
+    const newOutput = JSON.parse(solc.compile(JSON.stringify(newInput)));
+    st.ok(newOutput);
+
+    const errors = newOutput.errors;
+    st.equal(errors.length, 1);
+    st.equal(errors[0].errorCode, '6328');
```

Comment:
Has this been tested with a pre-release binary of the compiler? I'm mostly worried about the pre-release warning that is always issued. It could add an extra error here and solc-js test in Solidity repo would then fail once we merge this.

Overall, I think that it would be a good idea to create a temporary branch in the main repo, switch the solc-js test to this PR and see if it passes. In the past we used to have breakage sometimes after merging stuff in solc-js because of stuff like this.

Reply:
Right, I'll switch this to just search for that error instead, safer.

Reply:
Yep, just changed it to filter the errors by the error code, and ran it with a pre-release binary that I compiled locally.